### PR TITLE
Automate pausing / unpausing projects when finishing / linking new data

### LINKED ISSUE #2365: pause a project when active workflows are finished
We should automatically pause a project when all the active workflows are finished. Like wise we should unpause it when new data is linked / new workflows are added.

### PR DESCRIPTION
**Pull request by camallen**
Closes #2365 - Pause a project when it's completed all the actively linked data for all active workflows. Unpause it when new data is linked and reset the completness metrics  and counts when linking / unlinking new data.

# TODO: 
- [x] Unpause a project when linking new data, via the completeness worker
- [x] Investigate better unique locking constraints in sidekiq workers
- [x] Look at chaining workers or spacing temporally (simplest) to avoid race conditions between counters

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: app/controllers/api/v1/workflows_controller.rb [hunk 1]
```diff
@@ -82,13 +82,13 @@ def field_context
   # be easily located in specific controller actions
   def post_link_actions(workflow)
     if workflow.set_member_subjects.exists?
+      RefreshWorkflowStatusWorker.perform_async(workflow.id)
       case relation
-      when :retired_subjects, 'retired_subjects'
+      when /\Aretired_subjects\z/
```
Comment (marten) [inline app/controllers/api/v1/workflows_controller.rb]: Not sure how I feel about this. Wouldn't comparing with a regex be slower? What do you think about changing the `case` to be `case relation.to_s`

Comment (camallen) [reply app/controllers/api/v1/workflows_controller.rb]: Probably though benchmarks show it is negligible. I assumed reading the one regex is a little easier but the string comparison makes sense too.
```
       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000034)
string:  0.000000   0.000000   0.000000 (  0.000015)
       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000014)
string:  0.000000   0.000000   0.000000 (  0.000015)
       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000011)
string:  0.000000   0.000000   0.000000 (  0.000015)
.       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000025)
string:  0.000000   0.000000   0.000000 (  0.000009)
       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000013)
string:  0.000000   0.000000   0.000000 (  0.000009)
       user     system      total        real
regex:   0.000000   0.000000   0.000000 (  0.000011)
string:  0.000000   0.000000   0.000000 (  0.000011)
```

Comment (camallen) [reply app/controllers/api/v1/workflows_controller.rb]: here was the code
``` ruby
3.times do
  n = 10000
  Benchmark.bm do |x|
    x.report("regex: ") do
      case relation
      when /\Asubject_sets\z/ then true
      end
    end
    x.report("string:") do
      case relation.to_s
      when "subject_sets" then true
      end
    end
  end
end
```

**Comment by camallen** [comment]
This is ready to go but hasn't been approved, any takers to have a look?

**Comment by zwolf** [comment]
Does `workflow.set_member_subjects.exists?` give you enough information to un-finish a workflow? If I'm reading this correctly, then since calling the refresh worker is outside the case it'll un-finish a workflow regardless of what gets newly linked by the request.

**Comment by camallen** [comment]
>Does workflow.set_member_subjects.exists? give you enough information to un-finish a workflow? 

Yes, knowing that there is data linked is important however to speed up and simplify the logic here I've chosen to just call 1 worker. 

>If I'm reading this correctly, then since calling the refresh worker is outside the case it'll un-finish a workflow regardless of what gets newly linked by the request.

Yeah - the idea is to simply the controller logic and rely on 1 worker. The refresh SWS worker will run and unfinish the workflow (if finished). Then it'll recount the retired subjects and finish the workflow if needed. Finally, it'll update the project & workflow completeness metrics so the home page stats update as well.

As both retired and new subject links will impact these numbers we can rely on the order of unfinished, recount and update finish and then update metrics so this seemed a viable solution.

**Pull request merged**